### ISSUE TITLE
typeof const enum

### ISSUE THREAD

**Issue by jbondc** [microsoft/TypeScript#3440]
```
const enum Bar { a, b}
var c: typeof Bar.a; // 'const' enum can only be used in property or index access expressions or the right hand side of an import declaration or export assignment.
```

Should it be allowed?

**Comment by danquirk** [comment]
Seems reasonable. The error is clearly wrong as it's referencing the restrictions on const enums in value positions.

**Comment by mhegazy** [comment]
is not that just `Bar`? 

While I agree that this is an incorrect behavior, i do not see a real motivation to fix it.

**Comment by jbondc** [comment]
@mhegazy There's a subtle issue:
```
module thingsString {
   const var a = "a";
   const var b = "a";
}
var a: typeof thingsString

// If you re-write to a const enum that allows strings:

const enum thingsString {
   a = "a";
   b = "a";
}
var a: typeof thingsString
```
It's a simple fix:
```
                !((node.parent.kind === SyntaxKind.PropertyAccessExpression && (<PropertyAccessExpression>node.parent).expression === node) ||
                (node.parent.kind === SyntaxKind.ElementAccessExpression && (<ElementAccessExpression>node.parent).expression === node) ||
                ((node.kind === SyntaxKind.Identifier || node.kind === SyntaxKind.QualifiedName) && isInRightSideOfImportOrExportAssignment(<Identifier>node)) ||
				  isInTypeQuery(node))
				) {
```
Anyways let me if you change your mind.

**Comment by RyanCavanaugh** [comment]
Your first comment was about `typeof e.m`, but now you're talking about `typeof e`. Those are two very different things -- `typeof e.m` is a long way of writing `e`, but there's a reasonable use case to allow `typeof e`.

**Comment by mhegazy** [comment]
@jbondc feel free to send a PR for this if this is interesting to you. please limit the scope to skipping the error reporting if the use of const enum is in a type position.

**Comment by DanielRosenwasser** [comment]
> but there's a reasonable use case to allow `typeof e`

@RyanCavanaugh considering that const enums do not always have a runtime representation, could you explain what that might be?

**Comment by RyanCavanaugh** [comment]
In input positions (e.g. function parameters), `typeof E` seems equally valid whether `E` is const or not

**Comment by NoelAbrahams** [comment]
@mhegazy, can this be added to the next milestone? Seems an obvious bug that needs fixing.

Use case:

```ts
const enum Bar {
  good = 1,
  bad = 2,
  ugly = 4
}

// type Rating = 'good' | 'bad' | 'ugly';
type Rating = keyof typeof Bar; // Error: 'const' enums can only be used in property or index access expressions...
```

**Comment by mhegazy** [comment]
The bug is marked as `help wanted`. if we do get a PR for fixing it we will include it in the next release.

**Comment by NoelAbrahams** [comment]
Ah, okay. I was under the impression that only minor bugs are left to the community.

**Comment by mhegazy** [comment]
We do accept PRs for bugs and features regardless of complexity.

**Comment by beheh** [comment]
The `keyof typeof Bar` issue was reraised in #21031 and fixed in #21083 (which was released with 2.7). As such this issue can probably be closed.

**Comment by jakebailey** [comment]
Just looking at ancient issues; yeah, this one is fixed as `keyof typeof Bar` now works/
